File: app/src/main/java/ru/cool/lwjgl/objects/geometry/Geometry.java
package ru.cool.lwjgl.objects.geometry;

import ru.cool.lwjgl.objects.buffers.EBO;
import ru.cool.lwjgl.objects.buffers.VAO;
import ru.cool.lwjgl.objects.buffers.VBO;

import java.nio.FloatBuffer;
import java.nio.IntBuffer;

public abstract class Geometry {
    protected FloatBuffer vertices;
    protected IntBuffer indices;
    protected VAO vao;
    protected VBO<FloatBuffer> vbo;
    protected EBO ebo;

    public Geometry(){
        this.initMeshData();
        this.initGLObjects();
        this.initVertexAttributes();
    }

    /**
     * Initialization of mesh vertices and indexes
     */
    public abstract void initMeshData();

    /**
     * Adding vertex attributes. Before adding it, you need to bind vbo and ebo
     */
    public abstract void initVertexAttributes();

    public void initStandardAttributes(){
        vao.bindVAO();
        vbo.bindBuffer();
        ebo.bindBuffer();
        vao.enableVertexAttribute(0);
        ebo.setBufferData(indices);
        vao.addVertexAttribute(0, 3, 5 * Float.BYTES, 0);
        vao.enableVertexAttribute(1);
        vao.addVertexAttribute(1, 2, 5 * Float.BYTES, 3 * Float.BYTES);
        vbo.unbindBuffer();
        vao.unbindVAO();
    }
    public void initGLObjects(){
        vao = new VAO();
        vbo = new VBO<>();
        ebo = new EBO();

        vao.generateVAO();
        vao.bindVAO();
        vbo.generateBuffer();
        vbo.bindBuffer();
        vbo.setBufferData(vertices);
        ebo.generateBuffer();

        vbo.unbindBuffer();
        vao.unbindVAO();
    }
    public VAO getVao() {
        return vao;
    }

    public VBO<FloatBuffer> getVbo() {
        return vbo;
    }

    public EBO getEbo() {
        return ebo;
    }

    public FloatBuffer getVertices() {
        return vertices;
    }

    public IntBuffer getIndices() {
        return indices;
    }

}

File: kotlin_app/src/main/kotlin/ru/cool/lwjgl_kotlin/geometry/Geometry.kt
package ru.cool.lwjgl_kotlin.geometry

import org.lwjgl.opengl.GL30
import org.lwjgl.system.MemoryUtil
import ru.cool.lwjgl_kotlin.glObjects.EBO
import ru.cool.lwjgl_kotlin.glObjects.VAO
import ru.cool.lwjgl_kotlin.glObjects.VBO
import java.nio.FloatBuffer
import java.nio.IntBuffer

abstract class Geometry private constructor(
    ) {
    val vbo = VBO()
    val ebo = EBO()
    val vao = VAO()
    open var hasNormal = false;
    open var hasUV = false;
    open lateinit var vertices: FloatBuffer
    open lateinit var indices: IntBuffer

    constructor(
        vertices: FloatBuffer,
        indices: IntBuffer,
        hasUV: Boolean,
        hasNormal: Boolean
    ) : this() {
        this.vertices = vertices
        this.indices = indices
        this.hasNormal = hasNormal
        this.hasUV = hasUV
        init()
    }

    fun init(){
        vao.bind()
        vbo.bindBuffer()
        vbo.putDataToBuffer(vertices)
        ebo.bindBuffer()
        ebo.putDataToBuffer(indices)
        MemoryUtil.memFree(vertices)
        MemoryUtil.memFree(indices)
        vao.setVertexAttribute(
            0,
            3,
            if(hasUV && hasNormal)
                8 * Float.SIZE_BYTES
            else if (hasUV)
                5 * Float.SIZE_BYTES
            else if (hasNormal)
                6 * Float.SIZE_BYTES
            else
                3 * Float.SIZE_BYTES,
            0)
        if(hasNormal)
            vao.setVertexAttribute(
                1,
                3,
                8 * Float.SIZE_BYTES,
                3 * Float.SIZE_BYTES)
        if(hasUV)
            vao.setVertexAttribute(
                2,
                2,
                8 * Float.SIZE_BYTES,
                6 * Float.SIZE_BYTES
            )
        vao.unbind()
    }
}